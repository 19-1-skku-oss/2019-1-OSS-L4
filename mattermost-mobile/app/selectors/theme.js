// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {createSelector} from 'reselect';

import Preferences from 'mattermost-redux/constants/preferences';
import {getConfig} from 'mattermost-redux/selectors/entities/general';
import {getTheme} from 'mattermost-redux/selectors/entities/preferences';

export const getAllowedThemes = createSelector(
    getConfig,
    getTheme,
    (config) => {
        const allThemes = Object.keys(Preferences.THEMES).map((key) => ({
            ...Preferences.THEMES[key],
            key,
        }));
        let acceptableThemes = allThemes;
        const allowedThemeKeys = (config.AllowedThemes || '').split(',').filter(String);
        if (allowedThemeKeys.length) {
            acceptableThemes = allThemes.filter((theme) => allowedThemeKeys.includes(theme.key));
        }
        return acceptableThemes;
    }
);

export const getCustomTheme = createSelector(
    getConfig,
    getTheme,
    (config, activeTheme) => {
        if (config.AllowCustomThemes === 'true' && activeTheme.type === 'custom') {
            return {
                ...activeTheme,
                key: 'custom',
            };
        }
        return null;
    }
);