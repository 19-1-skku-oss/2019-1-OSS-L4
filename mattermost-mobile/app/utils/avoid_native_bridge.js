// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Platform} from 'react-native';

/**
 * avoidNativeBridge: a helper function that facilitates returning
 * a constant variable packaged from InitializationModule
 * or from a NativeModule which will make a call to the native bridge
 *
 * Currently only required for Android
 */
export default function avoidNativeBridge(runOptimized, optimized, fallback) {
    if (Platform.OS === 'android' && runOptimized()) {
        return optimized();
    }

    return fallback();
}
