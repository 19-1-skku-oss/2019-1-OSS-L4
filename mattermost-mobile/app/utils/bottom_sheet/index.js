// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import BottomSheet from './bottom_sheet';
export default BottomSheet;
