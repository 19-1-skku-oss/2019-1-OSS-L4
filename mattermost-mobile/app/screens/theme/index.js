// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {getTheme} from 'mattermost-redux/selectors/entities/preferences';
import {getCurrentUserId} from 'mattermost-redux/selectors/entities/users';
import {getCurrentTeamId} from 'mattermost-redux/selectors/entities/teams';
import {savePreferences} from 'mattermost-redux/actions/preferences';
import {getAllowedThemes, getCustomTheme} from 'app/selectors/theme';
import {isLandscape, isTablet} from 'app/selectors/device';

import Theme from './theme';

const mapStateToProps = (state) => ({
    allowedThemes: getAllowedThemes(state),
    customTheme: getCustomTheme(state),
    isLandscape: isLandscape(state),
    isTablet: isTablet(state),
    teamId: getCurrentTeamId(state),
    theme: getTheme(state),
    userId: getCurrentUserId(state),
});

const mapDispatchToProps = (dispatch) => ({
    actions: bindActionCreators({
        savePreferences,
    }, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(Theme);
