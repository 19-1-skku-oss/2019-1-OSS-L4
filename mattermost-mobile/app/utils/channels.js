// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Preferences} from 'mattermost-redux/constants';
import {getUserIdFromChannelName} from 'mattermost-redux/utils/channel_utils';

export function isDirectChannelVisible(userId, myPreferences, channel) {
    const channelId = getUserIdFromChannelName(userId, channel.name);
    const dm = myPreferences[`${Preferences.CATEGORY_DIRECT_CHANNEL_SHOW}--${channelId}`];
    return dm && dm.value === 'true';
}

export function isGroupChannelVisible(myPreferences, channel) {
    const gm = myPreferences[`${Preferences.CATEGORY_GROUP_CHANNEL_SHOW}--${channel.id}`];
    return gm && gm.value === 'true';
}
