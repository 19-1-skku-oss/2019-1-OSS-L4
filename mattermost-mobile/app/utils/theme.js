// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {StyleSheet} from 'react-native';

import * as ThemeUtils from 'mattermost-redux/utils/theme_utils';

export function makeStyleSheetFromTheme(getStyleFromTheme) {
    return ThemeUtils.makeStyleFromTheme((theme) => {
        return StyleSheet.create(getStyleFromTheme(theme));
    });
}

export const changeOpacity = ThemeUtils.changeOpacity;

export const blendColors = ThemeUtils.blendColors;

export function concatStyles(...styles) {
    return [].concat(styles);
}

export function setNavigatorStyles(navigator, theme) {
    navigator.setStyle({
        navBarTextColor: theme.sidebarHeaderTextColor,
        navBarBackgroundColor: theme.sidebarHeaderBg,
        navBarButtonColor: theme.sidebarHeaderTextColor,
        screenBackgroundColor: theme.centerChannelBg,
    });
}

export function isThemeSwitchingEnabled(state) {
    const {config} = state.entities.general;
    return config.EnableThemeSelection === 'true';
}
