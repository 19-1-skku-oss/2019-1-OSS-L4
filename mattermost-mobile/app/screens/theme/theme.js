// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {Text, View} from 'react-native';
import PropTypes from 'prop-types';
import {intlShape} from 'react-intl';

import StatusBar from 'app/components/status_bar';
import Section from 'app/screens/settings/section';
import SectionItem from 'app/screens/settings/section_item';
import ThemeTile from './theme_tile';
import FormattedText from 'app/components/formatted_text';

import {changeOpacity, makeStyleSheetFromTheme, setNavigatorStyles} from 'app/utils/theme';
import Preferences from 'mattermost-redux/constants/preferences';

const thumbnailImages = {
    default: require('assets/images/themes/mattermost.png'),
    organization: require('assets/images/themes/organization.png'),
    mattermostDark: require('assets/images/themes/mattermost_dark.png'),
    windows10: require('assets/images/themes/windows_dark.png'),
};

export default class Theme extends React.PureComponent {
    static propTypes = {
        actions: PropTypes.shape({
            savePreferences: PropTypes.func.isRequired,
        }).isRequired,
        allowedThemes: PropTypes.arrayOf(PropTypes.object),
        customTheme: PropTypes.object,
        isLandscape: PropTypes.bool.isRequired,
        isTablet: PropTypes.bool.isRequired,
        navigator: PropTypes.object.isRequired,
        teamId: PropTypes.string.isRequired,
        theme: PropTypes.object.isRequired,
        userId: PropTypes.string.isRequired,
    };

    static contextTypes = {
        intl: intlShape.isRequired,
    };

    state = {
        customTheme: null,
    };

    static getDerivedStateFromProps(props, state) {
        if (!state.customTheme && props.customTheme) {
            return {
                customTheme: props.customTheme,
            };
        }
        return null;
    }

    componentDidUpdate(prevProps) {
        if (prevProps.theme !== this.props.theme) {
            setNavigatorStyles(this.props.navigator, this.props.theme);
        }
    }

    setTheme = (key) => {
        const {userId, teamId, actions: {savePreferences}, allowedThemes} = this.props;
        const {customTheme} = this.state;
        const selectedTheme = allowedThemes.concat(customTheme).find((theme) => theme.key === key);

        savePreferences(userId, [{
            user_id: userId,
            category: Preferences.CATEGORY_THEME,
            name: teamId,
            value: JSON.stringify(selectedTheme),
        }]);
    }

    renderAllowedThemeTiles = () => {
        const {theme, allowedThemes, isLandscape, isTablet} = this.props;

        return allowedThemes.map((allowedTheme) => (
            <ThemeTile
                key={allowedTheme.key}
                label={(
                    <Text>
                        {allowedTheme.type}
                    </Text>
                )}
                action={this.setTheme}
                actionValue={allowedTheme.key}
                selected={allowedTheme.type.toLowerCase() === theme.type.toLowerCase()}
                theme={theme}
                imageSrc={thumbnailImages[allowedTheme.key]}
                isLandscape={isLandscape}
                isTablet={isTablet}
            />
        ));
    };

    renderCustomThemeRow = ({item}) => {
        const {theme} = this.props;

        return (
            <SectionItem
                label={(
                    <FormattedText
                        id='user.settings.display.custom_theme'
                        defaultMessage={'Custom Theme'}
                    />
                )}
                action={this.setTheme}
                actionType='select'
                actionValue={item.key}
                selected={item.type.toLowerCase() === theme.type.toLowerCase()}
                theme={theme}
            />
        );
    };

    render() {
        const {theme} = this.props;
        const {customTheme} = this.state;
        const style = getStyleSheet(theme);
        return (
            <View style={style.container}>
                <StatusBar/>
                <View style={style.wrapper}>
                    <View style={style.tilesContainer}>
                        {this.renderAllowedThemeTiles()}
                    </View>

                    {customTheme &&
                        <Section
                            disableHeader={true}
                            theme={theme}
                        >
                            {this.renderCustomThemeRow({item: customTheme})}
                        </Section>
                    }
                </View>
            </View>
        );
    }
}

const getStyleSheet = makeStyleSheetFromTheme((theme) => {
    return {
        container: {
            flex: 1,
            backgroundColor: theme.centerChannelBg,
        },
        wrapper: {
            backgroundColor: changeOpacity(theme.centerChannelColor, 0.06),
            flex: 1,
            paddingTop: 35,
        },
        tilesContainer: {
            marginBottom: 30,
            flexDirection: 'row',
            flexWrap: 'wrap',
            justifyContent: 'center',
            backgroundColor: theme.centerChannelBg,
            borderTopWidth: 1,
            borderBottomWidth: 1,
            borderTopColor: changeOpacity(theme.centerChannelColor, 0.1),
            borderBottomColor: changeOpacity(theme.centerChannelColor, 0.1),
        },
    };
});
