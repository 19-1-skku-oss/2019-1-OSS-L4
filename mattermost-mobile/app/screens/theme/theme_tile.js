// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import PropTypes from 'prop-types';
import {
    Dimensions,
    Image,
    TouchableOpacity,
    View,
} from 'react-native';

import {makeStyleSheetFromTheme} from 'app/utils/theme';

const {width: deviceWidth, height: deviceHeight} = Dimensions.get('window');

const checkmark = require('assets/images/themes/check.png');

const tilePadding = 8;

const ThemeTile = (props) => {
    const {
        action,
        actionValue,
        imageSrc,
        isLandscape,
        isTablet,
        label,
        selected,
        theme,
    } = props;

    const style = getStyleSheet(theme);

    const labelComponent = React.cloneElement(
        label,
        {style: style.label}
    );

    const tilesPerLine = isLandscape || isTablet ? 4 : 2;
    const fullWidth = isLandscape ? deviceHeight - 40 : deviceWidth;
    const layoutStyle = {
        container: {
            width: (fullWidth / tilesPerLine) - tilePadding,
        },
        thumbnail: {
            width: (fullWidth / tilesPerLine) - (tilePadding + 20),
        },
    };

    return (
        <TouchableOpacity
            style={[style.container, layoutStyle.container]}
            onPress={() => action(actionValue)}
        >
            <View style={[style.imageWrapper, layoutStyle.thumbnail]}>
                {imageSrc && (
                    <Image
                        source={imageSrc}
                        style={[
                            style.thumbnail,
                            layoutStyle.thumbnail,
                            selected ? style.selectedThumbnail : null,
                        ]}
                    />
                )}
                {selected && (
                    <Image
                        source={checkmark}
                        style={style.check}
                    />
                )}
            </View>
            {labelComponent}
        </TouchableOpacity>
    );
};

ThemeTile.propTypes = {
    action: PropTypes.func,
    actionValue: PropTypes.string,
    imageSrc: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.object,
    ]).isRequired,
    isLandscape: PropTypes.bool.isRequired,
    isTablet: PropTypes.bool.isRequired,
    label: PropTypes.node.isRequired,
    selected: PropTypes.bool,
    theme: PropTypes.object.isRequired,
};

ThemeTile.defaultProps = {
    action: () => true,
};

const getStyleSheet = makeStyleSheetFromTheme((theme) => {
    return {
        container: {
            flexDirection: 'column',
            padding: tilePadding,
            marginTop: 8,
        },
        imageWrapper: {
            position: 'relative',
            alignItems: 'flex-start',
            marginBottom: 12,
        },
        thumbnail: {
            resizeMode: 'stretch',
        },
        selectedThumbnail: {
            opacity: 0.5,
        },
        check: {
            position: 'absolute',
            right: 10,
            bottom: 10,
        },
        label: {
            color: theme.centerChannelColor,
            fontSize: 15,
        },
    };
});

export default ThemeTile;
