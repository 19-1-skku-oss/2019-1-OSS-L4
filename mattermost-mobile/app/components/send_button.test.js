// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import Preferences from 'mattermost-redux/constants/preferences';

import SendButton from 'app/components/send_button';
import {changeOpacity} from 'app/utils/theme';

describe('SendButton', () => {
    const baseProps = {
        theme: Preferences.THEMES.default,
        handleSendMessage: jest.fn(),
        disabled: false,
    };

    function getWrapper(props = {}) {
        return shallow(
            <SendButton
                {...baseProps}
                {...props}
            />
        );
    }

    test('should match snapshot', () => {
        const wrapper = getWrapper();

        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('should render theme backgroundColor', () => {
        const wrapper = getWrapper();

        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper.childAt(0)).toHaveStyle('backgroundColor', baseProps.theme.buttonBg);
    });

    test('should change theme backgroundColor to 0.3 opacity', () => {
        const wrapper = getWrapper({disabled: true});

        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper.childAt(0)).toHaveStyle('backgroundColor', changeOpacity(baseProps.theme.buttonBg, 0.3));
    });
});
