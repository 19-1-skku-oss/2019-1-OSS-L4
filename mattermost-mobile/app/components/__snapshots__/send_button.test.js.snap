// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendButton should change theme backgroundColor to 0.3 opacity 1`] = `
<View
  style={
    Object {
      "justifyContent": "flex-end",
      "paddingHorizontal": 5,
      "paddingVertical": 3,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#166de0",
          "borderRadius": 18,
          "height": 28,
          "justifyContent": "center",
          "width": 28,
        },
        Object {
          "backgroundColor": "rgba(22,109,224,0.3)",
        },
      ]
    }
  >
    <PaperPlane
      color="#ffffff"
      height={13}
      width={15}
    />
  </View>
</View>
`;

exports[`SendButton should match snapshot 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
  style={
    Object {
      "justifyContent": "flex-end",
      "paddingHorizontal": 5,
      "paddingVertical": 3,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#166de0",
        "borderRadius": 18,
        "height": 28,
        "justifyContent": "center",
        "width": 28,
      }
    }
  >
    <PaperPlane
      color="#ffffff"
      height={13}
      width={15}
    />
  </View>
</TouchableOpacity>
`;

exports[`SendButton should render theme backgroundColor 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
  style={
    Object {
      "justifyContent": "flex-end",
      "paddingHorizontal": 5,
      "paddingVertical": 3,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#166de0",
        "borderRadius": 18,
        "height": 28,
        "justifyContent": "center",
        "width": 28,
      }
    }
  >
    <PaperPlane
      color="#ffffff"
      height={13}
      width={15}
    />
  </View>
</TouchableOpacity>
`;
